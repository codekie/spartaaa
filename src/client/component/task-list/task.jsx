import React, { Component } from 'react';
import moment from 'moment';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Content from 'react-bulma-components/lib/components/content';
import Icon from 'react-bulma-components/lib/components/icon';
import Level from 'react-bulma-components/lib/components/level';
import Media from 'react-bulma-components/lib/components/media';
import Tag from 'react-bulma-components/lib/components/tag';

import TagList from '../tag-list';
import determineClassNames from './css-class-determinator';
import determineIcon from './icon-determinator';
import './task.sass';
import PropTypes from 'prop-types';

export default class Task extends Component {
    static propTypes = {
        task: PropTypes.object.isRequired
    };

    constructor(props) {
        super(props);
    }
    render() {
        const { task } = this.props;
        return (
            <Media className={`cmp-task ${ determineClassNames(task) }`}>
                { _createPriorityIndicator(task) }
                <Media.Item renderAs="figure" position="left">
                    <Icon className="is-medium fa-2x">
                        <FontAwesomeIcon icon={determineIcon(task)} className="task-icon" />
                    </Icon>
                </Media.Item>
                <Media.Item>
                    <Content>
                        <p className="sub-description">
                            <strong className="sub-title">[{ task.id }] { task.description }</strong>
                        </p>
                        <Level className="aux-bar-1">
                            <Level.Side align="left">
                            </Level.Side>
                        </Level>
                        <Level>
                            <Level.Side align="left" className="tags">
                                { _createDueItem(task) }
                                { _createProjectItem(task) }
                                <Level.Item><TagList tags={task.tags} /></Level.Item>
                            </Level.Side>
                        </Level>
                    </Content>
                    <Level breakpoint="mobile">
                        <Level.Side align="left">
                        </Level.Side>
                    </Level>
                </Media.Item>
                <Media.Item position="right">
                    <Content>
                        <small>{ task.urgency }</small>
                    </Content>
                </Media.Item>
            </Media>
        );
    }
}

function _createProjectItem(task) {
    if (!task.project) { return null; }
    return (
        <Level.Item>
            <Tag.Group gapless className="tag-project">
                <Tag color="dark">project</Tag>
                <Tag color="info">{ task.project }</Tag>
            </Tag.Group>
        </Level.Item>
    );
}

function _createDueItem(task) {
    if (!task.due) { return null; }
    const dueUntil = moment(task.due).fromNow(),
        classNameDue = task.due - Date.now() > 0 ? 'in-time' : 'overdue';
    return (
        <Level.Item>
            <Tag.Group gapless className="tag-due">
                <Tag color="dark">due</Tag>
                <Tag className={classNameDue}>{ dueUntil }</Tag>
            </Tag.Group>
        </Level.Item>
    );
}

function _createPriorityIndicator(task) {
    return <div className={`priority ${ task.priority ? `prio-${ task.priority }` : '' }`} />;
}
