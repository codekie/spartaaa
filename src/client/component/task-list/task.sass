@import '../../bulma-variables'

.cmp-task
  &.media + &.media
    margin-top: 0
  &.prio-high
    color: $red
    strong
      color: $red
  &.tag-next
    background-color: darken($purple, 30)
  &.status-active
    background-color: darken($purple, 15)
