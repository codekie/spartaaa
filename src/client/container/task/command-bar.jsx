// # IMPORTS

// 3rd-party
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
// App
import CommandBar from '../../component/task/command-bar.jsx';
import { extractImmutable } from '../../util';
import { createDispatchMapper } from '../../util/prop-mapper';
import PropTypes from 'prop-types';
import TaskStatus from '../../../comm/task-status';
import ActionType from '../../store/action-type';
import ActionCreator from '../../store/action-creators';

// # CONSTANTS

const MAP__DISPATCH_TO_PROPS = {
    [ActionType.activateTask]: ActionCreator[ActionType.activateTask],
    [ActionType.deactivateTask]: ActionCreator[ActionType.deactivateTask],
    [ActionType.finishTask]: ActionCreator[ActionType.finishTask],
    [ActionType.toggleNext]: ActionCreator[ActionType.toggleNext]
};


// # EXPORT PUBLIC API

export class ConnectedCommandBar extends PureComponent {
    static propTypes = {
        uuid: PropTypes.string.isRequired,

        activateTask: PropTypes.func.isRequired,
        deactivateTask: PropTypes.func.isRequired,
        toggleNext: PropTypes.func.isRequired,

        taskId: PropTypes.number,
        status: PropTypes.string,
        start: PropTypes.number,
        tags: PropTypes.array.isRequired
    };

    render() {
        const { taskId, tags, start, activateTask, deactivateTask, toggleNext, status } = this.props,
            isCompleted = status === TaskStatus.completed;
        return (<CommandBar taskId={taskId}
            isCompleted={isCompleted}
            isActive={!!start}
            activateTask={activateTask}
            deactivateTask={deactivateTask}
            tags={tags}
            toggleNext={toggleNext} />);
    }
}

export default connect(
    _mapStateToProps,
    createDispatchMapper(MAP__DISPATCH_TO_PROPS)
)(extractImmutable(ConnectedCommandBar));


function _mapStateToProps(state, ownProps) {
    const task = state.get('tasks').get('tasks').get(ownProps.uuid);
    return {
        taskId: task.get('id'),
        status: task.get('status'),
        start: task.get('start'),
        tags: task.get('tags').toJS()
    };
}
