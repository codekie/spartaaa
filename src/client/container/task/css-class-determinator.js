import TaskStatus from '../../../comm/task-status';

const PRIORITY_HIGH = 'H',
    TAG__BUG = 'bug',
    TAG__NEXT = 'next',
    CLASS_NAME__STATUS__ACTIVE = 'status-active',
    CLASS_NAME__STATUS__COMPLETED = 'status-completed',
    CLASS_NAME__PRIORITY__H = 'prio-high',
    CLASS_NAME__TAG__BUG = 'tag-bug',
    CLASS_NAME__TAG__NEXT = 'tag-next';

export default determineClassNames;

function determineClassNames(task) {
    const classNames = [],
        { tags = [], start, status, priority } = task;
    tags.includes(TAG__BUG) && classNames.push(CLASS_NAME__TAG__BUG);
    tags.includes(TAG__NEXT) && classNames.push(CLASS_NAME__TAG__NEXT);
    start && classNames.push(CLASS_NAME__STATUS__ACTIVE);
    status === TaskStatus.completed && classNames.push(CLASS_NAME__STATUS__COMPLETED);
    priority === PRIORITY_HIGH && classNames.push(CLASS_NAME__PRIORITY__H);
    return classNames.join(' ');
}
