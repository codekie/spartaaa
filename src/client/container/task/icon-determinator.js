import { faBug, faTasks } from '@fortawesome/free-solid-svg-icons';

const ICON__DEFAULT = faTasks,
    ICON__BUG = faBug,
    TAG__BUG = 'bug';

export default function determineIcon(task) {
    const tags = task.tags || [];
    if (tags.includes(TAG__BUG)) {
        return ICON__BUG;
    }
    return ICON__DEFAULT;
}
