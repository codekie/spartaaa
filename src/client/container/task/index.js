import ConnectedTask from './task.jsx';

export default ConnectedTask;
