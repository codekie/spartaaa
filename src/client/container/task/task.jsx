// # IMPORTS

// 3rd-party
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
// App
import Task from '../../component/task';
import { extractImmutable } from '../../util';
import { createDispatchMapper } from '../../util/prop-mapper';
import determineClassNames from './css-class-determinator';
import determineIcon from './icon-determinator';
import ActionType from '../../store/action-type';
import ActionCreator from '../../store/action-creators';

// # CONSTANTS

const MAP__DISPATCH_TO_PROPS = {
    [ActionType.activateTask]: ActionCreator[ActionType.activateTask],
    [ActionType.deactivateTask]: ActionCreator[ActionType.deactivateTask],
    [ActionType.finishTask]: ActionCreator[ActionType.finishTask],
    [ActionType.unfinishTask]: ActionCreator[ActionType.unfinishTask]
};


// # EXPORT PUBLIC API

class TaskContainer extends PureComponent {
    static propTypes = {
        uuid: PropTypes.string.isRequired
    };

    render() {
        return <Task {...this.props} />;
    }
}

export default connect(
    _mapStateToProps,
    createDispatchMapper(MAP__DISPATCH_TO_PROPS)
)(extractImmutable(TaskContainer));


function _mapStateToProps(state, ownProps) {
    const task = state.get('tasks').get('tasks').get(ownProps.uuid).toJS(),
        cssClassesString = determineClassNames(task),
        taskIcon = determineIcon(task);
    // Don't pass the `tags` property, since this will cause the component always to re-render, since the tags-array
    // always will be replaces with a new instance (and that even though it's not  even relevant to the
    // `Task`-component)
    delete task.tags;
    return {
        ...task,
        cssClassesString,
        taskIcon
    };
}
