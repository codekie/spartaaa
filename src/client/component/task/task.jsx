import React, { PureComponent } from 'react';
import moment from 'moment';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Content from 'react-bulma-components/lib/components/content';
import Icon from 'react-bulma-components/lib/components/icon';
import Level from 'react-bulma-components/lib/components/level';
import Media from 'react-bulma-components/lib/components/media';
import Tag from 'react-bulma-components/lib/components/tag';
import { faFolder, faClock, faCheck, faTimesCircle } from '@fortawesome/free-solid-svg-icons';

import CommandBar from './command-bar.jsx';
import ConnectedTaskTagList from '../../container/task-tag-list';
import './task.sass';
import PropTypes from 'prop-types';
import Button from 'react-bulma-components/lib/components/button';
import TaskStatus from '../../../comm/task-status';

export default class Task extends PureComponent {
    static propTypes = {
        // Functions
        activateTask: PropTypes.func.isRequired,
        deactivateTask: PropTypes.func.isRequired,
        finishTask: PropTypes.func.isRequired,
        unfinishTask: PropTypes.func.isRequired,

        // Raw-data
        description: PropTypes.string,
        due: PropTypes.number,
        id: PropTypes.number.isRequired,
        priority: PropTypes.string,
        project: PropTypes.string,
        start: PropTypes.number,
        status: PropTypes.string,
        urgency: PropTypes.number,
        uuid: PropTypes.string.isRequired,
        // Pre-processed data
        cssClassesString: PropTypes.string,
        iconTask: PropTypes.object
    };

    constructor(props) {
        super(props);
    }
    render() {
        const props = this.props,
            {
                uuid, id, description, due, project, urgency, cssClassesString, taskIcon, priority, start, status,
                activateTask, deactivateTask, finishTask, unfinishTask
            } = props,
            isCompleted = status === TaskStatus.completed;
        return (
            <Media className={`cmp-task ${ cssClassesString }`}>
                { _createPriorityIndicator(priority) }
                <Media.Item renderAs="figure" position="left">
                    <div>
                        <Icon className="is-medium fa-2x">
                            <FontAwesomeIcon icon={taskIcon} className="task-icon" />
                        </Icon>
                    </div>
                    <div>
                        { _createToggleStatusButton({ id, uuid, isCompleted, unfinishTask, finishTask }) }
                    </div>
                </Media.Item>
                <Media.Item>
                    <Content>
                        <p className="sub-description">
                            <strong className="sub-title"
                            >[{ id || uuid }] { description }</strong>
                        </p>
                        <Level>
                            <Level.Side align="left" className="tags">
                                { _createDueItem(due) }
                                { _createProjectItem(project) }
                                <Level.Item><ConnectedTaskTagList uuid={uuid} /></Level.Item>
                            </Level.Side>
                        </Level>
                        <Level className="aux-bar-1">
                            <Level.Side align="left">
                                <CommandBar taskId={id}
                                    isCompleted={isCompleted}
                                    isActive={!!start}
                                    activateTask={activateTask}
                                    deactivateTask={deactivateTask} />
                            </Level.Side>
                        </Level>
                    </Content>
                    <Level breakpoint="mobile">
                        <Level.Side align="left">
                        </Level.Side>
                    </Level>
                </Media.Item>
                <Media.Item position="right">
                    <Content>
                        <small>{ urgency }</small>
                    </Content>
                </Media.Item>
            </Media>
        );
    }
}

function _createProjectItem(project) {
    if (!project) { return null; }
    return (
        <Level.Item>
            <Tag.Group gapless className="tag-project tag-icon">
                <Tag color="dark">
                    <Icon className="is-small">
                        <FontAwesomeIcon icon={faFolder} />
                    </Icon>
                </Tag>
                <Tag color="info">{ project }</Tag>
            </Tag.Group>
        </Level.Item>
    );
}

function _createDueItem(due) {
    if (!due) { return null; }
    const dueUntil = moment(due).fromNow(),
        classNameDue = due - Date.now() > 0 ? 'in-time' : 'overdue';
    return (
        <Level.Item>
            <Tag.Group gapless className="tag-due tag-icon">
                <Tag color="dark">
                    <Icon className="is-small">
                        <FontAwesomeIcon icon={faClock} />
                    </Icon>
                </Tag>
                <Tag className={classNameDue}>{ dueUntil }</Tag>
            </Tag.Group>
        </Level.Item>
    );
}

function _createPriorityIndicator(priority) {
    return <div className={`priority ${ priority ? `prio-${ priority }` : '' }`} />;
}

function _createToggleStatusButton({ id, uuid, isCompleted, unfinishTask, finishTask } = {}) {
    return (
        <Button className="btn-check" onClick={() => {
            return isCompleted
                ? unfinishTask(uuid)
                : finishTask(id);
        }}>
            <Icon>
                <FontAwesomeIcon
                    icon={isCompleted
                        ? faTimesCircle
                        : faCheck}
                    className="check-icon" />
            </Icon>
        </Button>
    );
}
