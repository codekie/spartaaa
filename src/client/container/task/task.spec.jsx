import React from 'react';
import { shallow } from 'enzyme';
import { TaskContainer } from './task.jsx';

it('renders correctly', () => {
    const tree = shallow(<TaskContainer uuid="ABC" id={1} unfinishTask={jest.fn()} finishTask={jest.fn()}
        activateTask={jest.fn()} deactivateTask={jest.fn()} />
    ).debug();
    expect(tree).toMatchSnapshot();
});
