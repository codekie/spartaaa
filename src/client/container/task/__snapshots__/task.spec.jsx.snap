// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `"<Task uuid=\\"ABC\\" id={1} unfinishTask={[Function: mockConstructor]} finishTask={[Function: mockConstructor]} activateTask={[Function: mockConstructor]} deactivateTask={[Function: mockConstructor]} />"`;
